### PR TITLE
Feature/castelletto1/lai unit testing

### PR DESCRIPTION
**Pull request by castelletto1**
Fix unit tests on Lassen

### CODE REVIEW

File: src/coreComponents/unitTests/linearAlgebraTests/testDofManagerUtils.hpp [hunk 1]
```diff
@@ -490,7 +555,20 @@ void makeSparsityFlux( MeshLevel const * const mesh,
         }
       }
 
-      sparsity.insert( localDofIndex, localDofIndex, localValues );
+//      sparsity.insert( localDofIndex, localDofIndex, localValues );
```
Comment (rrsettgast) [inline src/coreComponents/unitTests/linearAlgebraTests/testDofManagerUtils.hpp:558]:
```suggestion
```

File: src/coreComponents/unitTests/linearAlgebraTests/testDofManagerUtils.hpp [hunk 2]
```diff
@@ -443,7 +495,20 @@ void makeSparsityMass( MeshLevel const * const mesh,
     {
       localDofIndex[c] = elemDofIndex[idx[0]][idx[1]][idx[2]] + c;
     }
-    sparsity.insert( localDofIndex, localDofIndex, localValues );
+//    sparsity.insert( localDofIndex, localDofIndex, localValues );
```
Comment (rrsettgast) [inline src/coreComponents/unitTests/linearAlgebraTests/testDofManagerUtils.hpp:498]:
```suggestion
```

File: src/coreComponents/unitTests/linearAlgebraTests/testDofManagerUtils.hpp [hunk 3]
```diff
@@ -407,8 +433,34 @@ void makeSparsityFEM_FVM( MeshLevel const * const mesh,
         }
       }
 
-      sparsity.insert( localNodeDofIndex, localElemDofIndex, localValues1 );
-      sparsity.insert( localElemDofIndex, localNodeDofIndex, localValues2 );
+//      sparsity.insert( localNodeDofIndex, localElemDofIndex, localValues1 );
+//      sparsity.insert( localElemDofIndex, localNodeDofIndex, localValues2 );
+
```
Comment (rrsettgast) [inline src/coreComponents/unitTests/linearAlgebraTests/testDofManagerUtils.hpp:438]:
```suggestion
```

File: src/coreComponents/unitTests/linearAlgebraTests/testDofManagerUtils.hpp [hunk 4]
```diff
@@ -343,8 +359,18 @@ void makeSparsityFEM( MeshLevel const * const mesh,
           localDofIndex[a * numComp + c] = nodeDofIndex[nodeMap[k][a]] + c;
         }
       }
-
-      sparsity.insert( localDofIndex, localDofIndex, localValues );
+//      sparsity.insertNonZero( localDofIndex, localDofIndex, localValues );
```
Comment (rrsettgast) [inline src/coreComponents/unitTests/linearAlgebraTests/testDofManagerUtils.hpp:362]:
```suggestion
```

File: src/coreComponents/unitTests/linearAlgebraTests/testDofManagerUtils.hpp [hunk 5]
```diff
@@ -275,31 +275,47 @@ void makeSparsityTPFA( MeshLevel const * const mesh,
   localIndex const numElem = faceToElem.size( 1 );
 
   array1d< globalIndex > localDofIndex( numElem * numComp );
+
   array2d< real64 > localValues( numElem * numComp, numElem * numComp );
   localValues.setValues< serialPolicy >( 1.0 );
 
   // Loop over faces and assemble TPFA-style "flux" contributions
-  forLocalObjects< DofManager::Location::Face >( mesh, regions, [&]( localIndex const kf )
+//  forLocalObjects< DofManager::Location::Face >( mesh, regions, [&]( localIndex const kf )
```
Comment (rrsettgast) [inline src/coreComponents/unitTests/linearAlgebraTests/testDofManagerUtils.hpp:283]:
```suggestion
```

File: src/coreComponents/unitTests/linearAlgebraTests/testDofManagerUtils.hpp [hunk 6]
```diff
@@ -275,31 +275,47 @@ void makeSparsityTPFA( MeshLevel const * const mesh,
...
+        if( localDofNumber < 0 || localDofNumber >= sparsity.numRows() )
+          continue;
+        sparsity.insertNonZeros( localDofNumber,
+                                 localDofIndex.data(),
+                                 localValues.data(),
+                                 count );
+      }
     }
-  } );
+  } //);
```
Comment (rrsettgast) [inline src/coreComponents/unitTests/linearAlgebraTests/testDofManagerUtils.hpp:318]:
```suggestion
  }
```

File: src/coreComponents/unitTests/linearAlgebraTests/testDofManager.cpp [hunk 7]
```diff
@@ -507,17 +509,22 @@ void DofManagerSparsityTest< LAI >::test( std::vector< FieldDesc > fields,
...
   }
 
+  CRSMatrix< real64, globalIndex > localPatternExpected( numLocalDof,
+                                                         dofManager.numGlobalDofs(),
+                                                         27 * numCompTotal );
   Matrix patternExpected;
-  patternExpected.createWithLocalSize( numLocalDof, numLocalDof, 27 * numCompTotal, MPI_COMM_GEOSX );
-  patternExpected.open();
+//  patternExpected.createWithLocalSize( numLocalDof, numLocalDof, 27 * numCompTotal, MPI_COMM_GEOSX );
+//  patternExpected.open();
```
Comment (rrsettgast) [inline src/coreComponents/unitTests/linearAlgebraTests/testDofManager.cpp:517]:
```suggestion
```

File: src/coreComponents/linearAlgebra/unitTests/testVectors.cpp [hunk 8]
```diff
@@ -564,6 +621,7 @@ REGISTER_TYPED_TEST_SUITE_P( VectorTest,
                              getMultipleValues,
                              setSingleValue,
                              addSingleValue,
+//                             setValues,
```
Comment (rrsettgast) [inline src/coreComponents/linearAlgebra/unitTests/testVectors.cpp:624]:
```suggestion
```

File: src/coreComponents/linearAlgebra/unitTests/testVectors.cpp [hunk 9]
```diff
@@ -367,7 +372,59 @@ TYPED_TEST_P( VectorTest, addSingleValue )
...
+//
+//  array1d< real64 > const valuesExtracted( localSize );
+//  x.extract( valuesExtracted );
+//  std::cout<<"breakpoint 3"<<std::endl;
+//
+//  forAll<serialPolicy>( localSize,
+//                        [=] GEOSX_HOST (localIndex const i )
+//  {
+//    EXPECT_EQ( valuesExtracted[i], valuesView[i] );
+//  });
```
Comment (rrsettgast) [inline src/coreComponents/linearAlgebra/unitTests/testVectors.cpp:424]:
```suggestion
```

File: src/coreComponents/linearAlgebra/unitTests/testVectors.cpp [hunk 10]
```diff
@@ -328,15 +330,18 @@ TYPED_TEST_P( VectorTest, setSingleValue )
 
   localIndex const localSize = 3;
 
+  std::cout<<"breakpoint1 "<<std::endl;
```
Comment (rrsettgast) [inline src/coreComponents/linearAlgebra/unitTests/testVectors.cpp:333]:
```suggestion
```

File: src/coreComponents/linearAlgebra/unitTests/testVectors.cpp [hunk 11]
```diff
@@ -328,15 +330,18 @@ TYPED_TEST_P( VectorTest, setSingleValue )
...
   Vector x;
   x.createWithLocalSize( localSize, MPI_COMM_GEOSX );
 
+  std::cout<<"breakpoint2 "<<std::endl;
```
Comment (rrsettgast) [inline src/coreComponents/linearAlgebra/unitTests/testVectors.cpp:337]:
```suggestion
```

File: src/coreComponents/linearAlgebra/unitTests/testVectors.cpp [hunk 12]
```diff
@@ -328,15 +330,18 @@ TYPED_TEST_P( VectorTest, setSingleValue )
...
   x.open();
   for( localIndex i = 0; i < localSize; ++i )
   {
     x.set( x.getGlobalRowID( i ), x.ilower() + i );
   }
   x.close();
+  std::cout<<"breakpoint3 "<<std::endl;
```
Comment (rrsettgast) [inline src/coreComponents/linearAlgebra/unitTests/testVectors.cpp:344]:
```suggestion
```

File: src/coreComponents/linearAlgebra/unitTests/testLinearAlgebraUtils.hpp [hunk 13]
```diff
@@ -317,52 +362,171 @@ void compute2DElasticityOperator( MPI_Comm const comm,
...
+//      arraySlice1d< globalIndex const > const columns = matrix.getColumns( iRow );
+//      arraySlice1d< real64 > const entries = matrix.getEntries( iRow );
+//      localIndex const numEntries = matrix.numNonZeros( iRow );
+//      GEOSX_LOG_RANK_VAR( iRow );
+//      GEOSX_LOG_RANK_VAR( numEntries );
+//      GEOSX_LOG_RANK_VAR( columns );
+//      GEOSX_LOG_RANK_VAR( entries );
+//    }
+//  }
+//  //////////////////
```
Comment (rrsettgast) [inline src/coreComponents/linearAlgebra/unitTests/testLinearAlgebraUtils.hpp:492]:
```suggestion

```

File: src/coreComponents/linearAlgebra/interfaces/direct/SuiteSparse.cpp [hunk 14]
```diff
@@ -59,22 +60,22 @@ using SSlong = SuiteSparse_long;
...
-static_assert( sizeof( SSlong ) == sizeof( globalIndex ),
-               "SuiteSparse Int and geosx::globalIndex must have the same size" );
-
-static_assert( std::is_signed< SSlong >::value == std::is_signed< globalIndex >::value,
-               "SuiteSparse Int and geosx::globalIndex must both be signed or unsigned" );
+//static_assert( sizeof( SSlong ) == sizeof( globalIndex ),
+//               "SuiteSparse Int and geosx::globalIndex must have the same size" );
+//
+//static_assert( std::is_signed< SSlong >::value == std::is_signed< globalIndex >::value,
+//               "SuiteSparse Int and geosx::globalIndex must both be signed or unsigned" );
```
Comment (francoishamon) [inline src/coreComponents/linearAlgebra/interfaces/direct/SuiteSparse.cpp:67]:
```suggestion
```
(if we don't want to keep the checks)

File: src/coreComponents/linearAlgebra/interfaces/direct/SuperLUDist.cpp [hunk 15]
```diff
@@ -141,7 +144,13 @@ struct SuperLUDistData
 
   ~SuperLUDistData()
   {
-    Destroy_CompRowLoc_Matrix_dist( &mat );
+//    Destroy_CompRowLoc_Matrix_dist( &mat ); //// <-- copy body
+    NRformat_loc *Astore = (NRformat_loc *)mat.Store;
+//    SUPERLU_FREE(Astore->rowptr);
+//    SUPERLU_FREE(Astore->colind);
+//    SUPERLU_FREE(Astore->nzval);
```
Comment (francoishamon) [inline src/coreComponents/linearAlgebra/interfaces/direct/SuperLUDist.cpp:151]:
```suggestion
    NRformat_loc *Astore = (NRformat_loc *)mat.Store;
```

File: src/coreComponents/linearAlgebra/interfaces/hypre/HypreVector.cpp [hunk 16]
```diff
@@ -207,10 +235,20 @@ void HypreVector::set( globalIndex const * globalIndices,
...
   GEOSX_LAI_ASSERT_GE( *std::min_element( globalIndices, globalIndices + size ), ilower() );
   GEOSX_LAI_ASSERT_GE( iupper(), getLocalRowID( *std::max_element( globalIndices, globalIndices + size ) ) );
-  GEOSX_LAI_CHECK_ERROR( HYPRE_IJVectorSetValues( m_ij_vector,
-                                                  LvArray::integerConversion< HYPRE_Int >( size ),
-                                                  hypre::toHypreBigInt( globalIndices ),
-                                                  values ) );
+//  GEOSX_LAI_CHECK_ERROR( HYPRE_IJVectorSetValues( m_ij_vector,
+//                                                  LvArray::integerConversion< HYPRE_Int >( size ),
+//                                                  hypre::toHypreBigInt( globalIndices ),
+//                                                  values ) );
```
Comment (francoishamon) [inline src/coreComponents/linearAlgebra/interfaces/hypre/HypreVector.cpp:241]:
```suggestion
```

File: src/coreComponents/linearAlgebra/unitTests/testLinearAlgebraUtils.hpp [hunk 17]
```diff
@@ -317,52 +362,156 @@ void compute2DElasticityOperator( MPI_Comm const comm,
...
 
-  // Close the matrix
-  elasticity2D.close();
+  // Construct the 2D elasticity (plane strain) operator
+  elasticity2D.create( matrix.toViewConst(), comm );
+
+//  //////////////////////////
+//  elasticity2D.write("mat.mtx");
+//  GEOSX_ERROR("Stop here");
+//  //////////////////////////
```
Comment (francoishamon) [inline src/coreComponents/linearAlgebra/unitTests/testLinearAlgebraUtils.hpp:484]:
```suggestion
```

File: src/coreComponents/unitTests/linearAlgebraTests/testDofManager.cpp [hunk 18]
```diff
@@ -533,12 +538,16 @@ void DofManagerSparsityTest< LAI >::test( std::vector< FieldDesc > fields,
...
                            getRegions( mesh, c.regions ),
+                           dofManager.rankOffset(),
                            f1.components,
                            f2.components,
-                           patternExpected );
+                           localPatternExpected );
   }
-  patternExpected.close();
+  patternExpected.create( localPatternExpected.toViewConst(), MPI_COMM_GEOSX );
+//  patternExpected.close();
```
Comment (francoishamon) [inline src/coreComponents/unitTests/linearAlgebraTests/testDofManager.cpp:547]:
```suggestion
```

**Review by francoishamon — APPROVED** [review]
Thanks a lot for fixing the unit tests @castelletto1 @rrsettgast! I am only able to make cosmetic comments to help the cleanup in these parts of the code, but the changes look very good to me.

File: src/coreComponents/linearAlgebra/interfaces/hypre/HypreExport.cpp [hunk 19]
```diff
@@ -65,114 +66,191 @@ void HypreExport::exportCRS( HypreMatrix const & mat,
...
     int const rank = MpiWrapper::commRank( vec.getComm() );
-    GEOSX_ERROR_IF( rank == m_targetRank && !values, "HypreExport: must pass non-null pointers on the target rank" );
 
     hypre_Vector * const localVector = hypre_ParVectorToVectorAll( vec.unwrapped() );
     if( rank == m_targetRank )
     {
       HYPRE_Real const * const data = hypre_VectorData( localVector );
-      std::copy( data, data + vec.globalSize(), values );
+#if defined(GEOSX_USE_HYPRE_CUDA)
+      cudaMemcpy( values.data(), data, vec.globalSize() * sizeof( HYPRE_Real ), cudaMemcpyDeviceToHost );
```
Comment (TotoGaz) [inline src/coreComponents/linearAlgebra/interfaces/hypre/HypreExport.cpp:164]: Is there a follow-up in order to keep our arrays/coda "CUDA agnostic"?

File: src/coreComponents/linearAlgebra/interfaces/hypre/HypreExport.cpp [hunk 20]
```diff
@@ -49,132 +50,197 @@ HypreExport::~HypreExport()
...
-      HYPRE_ParVector parVector;
-      GEOSX_LAI_CHECK_ERROR( HYPRE_VectorToParVector( m_subComm,
-                                                      localVector,
-                                                      partitioning,
-                                                      &parVector ) );
-
-
-      // copy parVector local part over to the output vector
-      HYPRE_Real * const parVectorData = hypre_VectorData( hypre_ParVectorLocalVector( parVector ) );
       std::copy( parVectorData, parVectorData + vec.localSize(), vec.extractLocalVector() );
```
Comment (castelletto1) [inline src/coreComponents/linearAlgebra/interfaces/hypre/HypreExport.cpp:220]:
```suggestion
      copyOrTransform( parVectorData, parVectorData + vec.localSize(), vec.extractLocalVector() );
```

Comment (klevzoff) [reply src/coreComponents/linearAlgebra/interfaces/hypre/HypreExport.cpp:220]: Sure, why not. In this case types are guaranteed to be the same by `static_assert` so it should always call the copy version, but for uniformity we can do this.

**Pull request merged**